Add vertical writing mode support to `RenderMenuList`

pxlcoder opened the pull request:
#### 6caf36aefca6b4a87410238aa7b3bf7e6cdad668
<pre>
Add vertical writing mode support to `RenderMenuList`
<a href="https://bugs.webkit.org/show_bug.cgi?id=264267">https://bugs.webkit.org/show_bug.cgi?id=264267</a>
<a href="https://rdar.apple.com/118007185">rdar://118007185</a>

Reviewed by Alan Baradlay and Tim Nguyen.

Update `RenderMenuList` to use logical properties and values where applicable.

Add WPTs for `&lt;select&gt;` with a vertical writing mode, similar to other controls.
Note that the computed style test will only pass in WebKit once `&lt;select&gt;` has
been marked testable.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-horizontal.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-horizontal.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-vertical.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-vertical.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-horizontal.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-horizontal.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-lr.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-lr.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-rl.optional-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-rl.optional.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style-expected.txt: Added.
* Source/WebCore/rendering/RenderMenuList.cpp:
(WebCore::RenderMenuList::adjustInnerStyle):
(RenderMenuList::computeIntrinsicLogicalWidths const):
(RenderMenuList::computePreferredLogicalWidths):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setMarginBefore):
(WebCore::RenderStyle::setMarginAfter):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setLogicalMinWidth):
(WebCore::RenderStyle::setLogicalMaxWidth):

Canonical link: <a href="https://commits.webkit.org/270309@main">https://commits.webkit.org/270309@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aef65cc4703858c1a75fb70c59cae6ba6422a3c3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25104 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3645 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27219 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23039 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25372 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1081 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23300 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25347 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2673 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21668 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27798 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2372 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28739 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22904 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22958 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26555 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2316 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/612 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path-svg-content/clip-path-content-use-007.svg (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3662 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22352 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6020 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2755 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2651 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical.optional.html:
I would add vertical-rl variants as well (for both tests)

pxlcoder replied:
I can't consistently assert the behavior for a `vertical-lr` and `vertical-rl` due to browser-specific UI differences. In WebKit, the two would appear the same, since we use a double-sided arrow as our indicator. Chrome and Firefox use a single arrow, and consequently, the direction changes depending on `rl` or `lr`.

This means it's not possible to write a match/mismatch test that would pass everywhere.

pxlcoder replied:
Actually, I could probably add a match test for just the `appearance: none` case. Will try that.

alanbaradlay commented on Source/WebCore/rendering/RenderMenuList.cpp:
```diff
@@ -116,21 +116,36 @@ void RenderMenuList::setInnerRenderer(RenderBlock& innerRenderer)
+    if (isHorizontalWritingMode)
+        innerStyle.setMinWidth(Length(0, LengthType::Fixed));
+    else
+        innerStyle.setMinHeight(Length(0, LengthType::Fixed));
```
it looks like we have RenderStyle::logicalMin/MaxWidth(), maybe we should have a setLogicalMin/MaxWidth too?

pxlcoder replied:
Added.

alanbaradlay commented on Source/WebCore/rendering/RenderMenuList.cpp:
```diff
@@ -116,21 +116,36 @@ void RenderMenuList::setInnerRenderer(RenderBlock& innerRenderer)
+        } else {
+            innerStyle.setMarginLeft(Length());
+            innerStyle.setMarginRight(Length());
+        }
```
same as above, we already have RenderStyle::setMarginStart/End, we are just missing setMarginBefore/After

pxlcoder replied:
Added.

alanbaradlay commented on Source/WebCore/rendering/RenderMenuList.cpp:
```diff
@@ -322,12 +337,15 @@ LayoutRect RenderMenuList::controlClipRect(const LayoutPoint& additionalOffset)
     maxLogicalWidth = shouldApplySizeContainment() ? theme().minimumMenuListSize(style()) : std::max(m_optionsWidth, theme().minimumMenuListSize(style()));
-    maxLogicalWidth += m_innerBlock->paddingLeft() + m_innerBlock->paddingRight();
+    if (style().isHorizontalWritingMode())
+        maxLogicalWidth += m_innerBlock->paddingLeft() + m_innerBlock->paddingRight();
```
paddingStart() + paddingEnd() does not work?

pxlcoder replied:
Adopted.

alanbaradlay commented on LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-rl.optional.html:
```diff
@@ -0,0 +1,20 @@
+    <option>Option 3</option>
+    <option>Option 4</option>
+    <option>Option 5</option>
+</select>
```
I'd also test cases where the formatting context root has the writing mode set and not the select element itself. I think that's a more common use case. Also you can add some tests where you mix them e.g. vertical root with horizontal select and vice versa.

pxlcoder replied:
Discussed offline, there are already more general tests for the inheritance behavior in `css/css-writing-modes`, so we will not be adding more for specific controls.

pxlcoder committed:
Add vertical writing mode support to `RenderMenuList`
https://bugs.webkit.org/show_bug.cgi?id=264267
rdar://118007185

Reviewed by Alan Baradlay and Tim Nguyen.

Update `RenderMenuList` to use logical properties and values where applicable.

Add WPTs for `<select>` with a vertical writing mode, similar to other controls.
Note that the computed style test will only pass in WebKit once `<select>` has
been marked testable.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-horizontal.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-horizontal.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-vertical.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-vertical.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-horizontal.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-horizontal.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-lr.optional-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-lr.optional.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-rl.optional-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-none-vertical-rl.optional.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-appearance-native-computed-style-expected.txt: Added.
* Source/WebCore/rendering/RenderMenuList.cpp:
(WebCore::RenderMenuList::adjustInnerStyle):
(RenderMenuList::computeIntrinsicLogicalWidths const):
(RenderMenuList::computePreferredLogicalWidths):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setMarginBefore):
(WebCore::RenderStyle::setMarginAfter):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setLogicalMinWidth):
(WebCore::RenderStyle::setLogicalMaxWidth):

Canonical link: https://commits.webkit.org/270309@main
